test/jdk/javax/swing/text/html/CSS/bug4764897.java

```diff
@@ -0,0 +1,109 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+import javax.swing.JEditorPane;
+import javax.swing.JFrame;
+import javax.swing.JTextPane;
+import javax.swing.SwingUtilities;
+import javax.swing.text.View;
+import java.awt.Robot;
+import java.awt.Shape;
+
+/*
+ * @test
+ * @bug 4764897
+ * @key headful
+ * @summary Tests if text and borders in HTML table doesn't run over right edge
+ */
+
+public class bug4764897 {
+    private static JEditorPane pane;
+    private static JFrame frame;
+    private static volatile boolean passed = false;
+
+    public static void main(String[] args) throws Exception {
+        try {
+            Robot robot = new Robot();
+
+            SwingUtilities.invokeAndWait(bug4764897::createAndShowUI);
+            robot.waitForIdle();
+            robot.delay(500);
+
+            SwingUtilities.invokeAndWait(bug4764897::testUI);
+            robot.waitForIdle();
+            robot.delay(200);
+
+            if (!passed) {
+                throw new RuntimeException("Test failed!!" +
+                        " Borders of HTML table not rendered correctly");
+            }
+        } finally {
+            SwingUtilities.invokeAndWait(() -> {
+                if (frame != null) {
+                    frame.dispose();
+                }
+            });
+        }
+    }
+
+    public static void createAndShowUI() {
+        pane = new JTextPane();
+        pane.setContentType("text/html");
+        pane.setText("<html>"
+                + "<html><head><table border>"
+                + "<tr><td width=20%>one</td><td width=20%>&nbsp;</td>"
+                + "<td width=20%>ThisIsAnExtraWideWord</td><td width=40%>"
+                + "This is text that won't get displayed correctly.</td></tr>"
+                + "</table></body></html>");
+
+        frame = new JFrame("Table Border Test");
+        frame.getContentPane().add(pane);
+        frame.setSize(600,200);
+        frame.setVisible(true);
+    }
+
+    private static void testUI() {
+        Shape r = pane.getBounds();
+        View v = pane.getUI().getRootView(pane);
+        int tableWidth = 0;
+        int cellsWidth = 0;
+        while (!(v instanceof javax.swing.text.html.ParagraphView)) {
+            int n = v.getViewCount();
+            Shape sh = v.getChildAllocation(n - 1,  r);
+            String viewName = v.getClass().getName();
+            if (viewName.endsWith("TableView")) {
+                tableWidth = r.getBounds().width;
+            }
+
+            if (viewName.endsWith("CellView")) {
+                cellsWidth = r.getBounds().x + r.getBounds().width;
```

Comment:
This looks weird to me: why does `cellWidth` include the `x` coordinate in its width? Perhaps, the variable is named incorrectly.

It looks we care about the last cell in the table only, and it is to be positioned so that it fits inside the table. In other words, the last cell is within table bounds, that's why `x + width` is used.

Reply:
Looks like a valid explanation as to why `x + width` is used.